### PR TITLE
Update types to fix #13

### LINKED ISSUE #13: Possible incorrect typespec
I am setting up some `@spec` contracts in my application and found what I think is an incorrect typespec in `JsonApiClient.Document`.

In `JsonApiClient.Document`, the `data` attribute [is listed as `[JsonApiClient.ResourceIdentifier.t()] | JsonApiClient.ResourceIdentifier.t() | nil`](https://github.com/Decisiv/json_api_client/blob/master/lib/json_api_client/document.ex#L143), but I believe it should also include `JsonApiClient.Resource.t()`, like so: `[JsonApiClient.Resource.t()] | [JsonApiClient.ResourceIdentifier.t()] | JsonApiClient.Resource.t() | JsonApiClient.ResourceIdentifier.t() | nil`.

Similarly, I think that `included` should also allow `JsonApiClient.Resource.t()`, [but doesn't](https://github.com/Decisiv/json_api_client/blob/master/lib/json_api_client/document.ex#L146).

I'd be happy to work on a change, but I wanted to verify that I am thinking about this correctly first.

### PR DESCRIPTION
**Pull request by amiel**
#### What does this PR do?

This PR adjusts some type specs and function specs to fix issues running dialyzer with code that uses `JsonApiClient`.

All of these changes affect specs only, which means there are no functional changes. These changes fix issues using dialyzer in my project.

* `%JsonApiClient.Document{data: }` now accepts `Resource.t() | [Resource.t()] | ResourceIdentifier.t() | [ResourceIdentifier.t()] | nil`, where before it would not accept the `Resource` case

  This allows writing a function that expects to parse a list of resources from a document response.

* A new type `JsonApiClient.Document.data()`, is a shortcut for `Resource.t() | [Resource.t()] | ResourceIdentifier.t() | [ResourceIdentifier.t()] | nil`

* `%JsonApiClient.Document{included: }` now accepts both `[ResourceIdentifier.t()]` and `[Resource.t()]`, where before it only accepted the `ResourceIdentifier` case

* `JsonApiClient.Request.resource/2` now accepts any of the options for the new `JsonApiClient.Document.data()` type

  This allows adding a list of resources or a list of resource identifiers to a request, where this was impossible before (with dialyzer checks).

  Also, the type for `%JsonApiClilent.Request{resource: }` is updated to accommodate this change.

* `JsonApiClient.Request.method/2` is updated to only take `http_method`s, where before it would only accept any `atom`

  I had a bug in my code where I used the method `:POST` instead of `:post`. JsonApiClient only adds the `resource` to the request if the method is a lowercase atom (like `:post`). This changes causes dialyzer to fail for my buggy code.

* `JsonApiClient.Request.path/2` can take a `ResourceIdentifier.t()` as well as a `Resource.t()` since all it needs is the `type` and `id`

#### Why?

Some type specs and function specs were too restrictive, where using `JsonApiClient` in a way that was supported would fail dialyzer checks. Another was not restrictive enough (`JsonApiClient.Request.method`), allowing incorrect code to pass dialyzer checks. These changes should help dialyzer prevent issues with client code.

#### What are the relevant tickets?

This resolves #13 

#### Other Questions:
- [x] Did I run this locally?
- [x] Did I run the whole test suite?

  Green for me.

- [ ] Did I add tests to cover the change?

  There are no functional changes, so I don't think there are any tests to write.

- [x] Did I run dialyzer?

  Yes, and I ran dialyzer in my project against these changes and they resolved my issues.

- [ ] Any additional deploy/release considerations?

  Not that I know of. Is there anything to be done to release this to hex? What about hexdocs?

- [ ] Does this PR cause necessary updates to the FAQ / documentation?

  I'm assuming hexdocs will pick up the changes automatically. Otherwise, I don't know; is there anything else I need to change?

### CODE REVIEW

File: lib/json_api_client/document.ex [hunk 1]
```diff
@@ -83,6 +83,8 @@ defmodule JsonApiClient.ResourceIdentifier do
     meta: map | nil
   }
   defstruct id: nil, type: nil, meta: nil
+
+  @type t_or_list :: t() | [t()]
```
Comment (amiel) [inline lib/json_api_client/document.ex]: I'd be happy to rename this if you have a better suggestion.

File: lib/json_api_client/document.ex [hunk 2]
```diff
@@ -134,16 +138,19 @@ defmodule JsonApiClient.JsonApi do
 end
 
 defmodule JsonApiClient.Document do
+
+  @type data :: JsonApiClient.Resource.t_or_list() | JsonApiClient.ResourceIdentifier.t_or_list() | nil
```
Comment (amiel) [inline lib/json_api_client/document.ex]: I also found this subtype useful within my project.

File: lib/json_api_client/request.ex [hunk 3]
```diff
@@ -65,7 +66,7 @@ defmodule JsonApiClient.Request do
   end
 
   @doc "Associate a path with this request"
-  @spec path(req :: Request.t(), Resource.t() | name) :: Request.t()
+  @spec path(req :: Request.t(), Resource.t() | ResourceIdentifier.t() | name) :: Request.t()
```
Comment (amiel) [inline lib/json_api_client/request.ex]: This is the only change here that I didn't need, but it seemed reasonable to me.

**Comment by amiel** [comment]
@tapajos I know you're busy. I appreciate any time you're willing to put in to review and releasing this.

Thanks so much :D

**Comment by tapajos** [comment]
Version `3.1.1` released.

**Comment by amiel** [comment]
@tapajos Thank you very much! 🍻

**Pull request merged**